File: leetcode/src/main/java/com/giwankim/leetcode/JumpGameII.java
package com.giwankim.leetcode;

public class JumpGameII {
  /**
   * @implNote Time {@code O(n)}, space {@code O(1)}.
   */
  public int jump(int[] nums) {
    int jumps = 0;
    int right = 0;
    int farthest = 0;
    for (int i = 0; i + 1 < nums.length; i++) {
      farthest = Math.max(farthest, i + nums[i]);
      if (i == right) {
        jumps += 1;
        right = farthest;
      }
    }
    return jumps;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/JumpGameII.kt
package com.giwankim.leetcode

class JumpGameII {
    fun jump(nums: IntArray): Int {
        var result = 0
        var farthest = 0
        var end = 0
        for (i in 0 until nums.size - 1) {
            farthest = maxOf(farthest, i + nums[i])
            if (i == end) {
                result += 1
                end = farthest
            }
        }
        return result
    }
}
